include/seqan/bam_io/write_bam.h

```diff
@@ -54,12 +54,41 @@ void write(TTarget & target,
            BamIOContext<TNameStore, TNameStoreCache, TStorageSpec> & context,
            Bam const & /*tag*/)
 {
+    String<bool> writtenSeqInfos;
+    resize(writtenSeqInfos, length(contigNames(context)), false);
+
     write(target, "BAM\1");
     clear(context.buffer);
 
     // Create text of header.
+    size_t globalRefId = 0;
     for (unsigned i = 0; i < length(header); ++i)
-        write(context.buffer, header[i], context, Sam());
+    {
+        BamHeaderRecord const & record = header[i];
+        if (record.type == BAM_HEADER_REFERENCE)
+            for (unsigned j = 0; j < length(record.tags); ++j)
+                if (record.tags[j].i1 == "SN")
+                {
+                    if (getIdByName(globalRefId, contigNamesCache(context), record.tags[j].i2))
+                        writtenSeqInfos[globalRefId] = true;
+                    break;
+                }
+
+        write(context.buffer, record, context, Sam());
+    }
+
+    // Write missing @SQ header records.
+    SEQAN_ASSERT_LEQ(length(contigLengths(context)), length(contigNames(context)));
+    for (unsigned i = 0; i < length(contigLengths(context)); ++i)
+    {
+        if (writtenSeqInfos[i])
+            continue;
+        write(context.buffer, "@SQ\tSN:");
+        write(context.buffer, contigNames(context)[i]);
+        write(context.buffer, "\tLN:");
+        appendNumber(context.buffer, contigLengths(context)[i]);
+        writeValue(context.buffer, '\n');
+    }
```

Comment:
```suggestion
    write(target, "BAM\1");
    clear(context.buffer);
    
    write(target, header, context, Sam());
```
Can you check if this works? 
I noticed you just copied from the SAM header. Seems very redundant.

Reply:
Yup, also works. Though it has to be `write(context.buffer` instead of `write(target`